### ISSUE TITLE
plac does not work with console_scripts in setup.py

### ISSUE THREAD

**Issue by copernico** [micheles/plac#31]
I'm writing a program that I want to distribute with a setup.py that contains the following lines:
```
entry_points={
        'console_scripts': [
            'git-explorer=git_explorer.cli:main',
            'git-explorer-similar=git_explorer.find_similar:main',
        ],
    },
```

the `find_similar.py` contains this:

```
if __name__ == '__main__':
    plac.call(main)
```

And the `main` function declaration looks like this:
```
@plac.annotations(
    repo_url=("Repository", "positional", None, str, None, 'REPOSITORY'),
    commit_id=("Commit", "positional", None, str, None, 'COMMIT'),
    limit=("Only search among the LIMIT latest commits (tip: supports negavive numbers)", "option", 'l', int, None, 'LIMIT'),
    fast=("Skip fetching if local repository exists", "flag", 's', bool, False),
    git_cache= ("Git repository dir", "option", 'g', str, GIT_CACHE, 'GIT_CACHE'),
    verbose=("Verbose", "flag", 'v', bool),
    )
def main(repo_url, commit_id, limit=0, fast=False, git_cache=GIT_CACHE, verbose=False):
.....
.....
```

Now, if I install the package and try to use the console script `git-explorer-similar` I get the following:

```
Traceback (most recent call last):
  File "/home/xxxxxx/.local/share/virtualenvs/git_explorer-Hd8x6_Ku/bin/git-explorer-similar", line 11, in <module>
    load_entry_point('git-explorer', 'console_scripts', 'git-explorer-similar')()
TypeError: main() missing 2 required positional arguments: 'repo_url' and 'commit_id'
```

What is the correct way to use plac in combination with the `console_scripts` directive of a standard `setup.py`? I could not find anything in the otherwise excellent documentation...

**Comment by micheles** [comment]
From the error message it seems that console_scripts is not passing the required arguments to your `main()` function. You should study how console_scripts works and how it gets the argument. It could be that the problem has nothing to do with plac.

**Comment by copernico** [comment]
Hi Michele, thanks for your prompt reply. Actually console_scripts expects to call an entry-point function that takes **no arguments**, according to: https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point
However, this conflicts with the **@plac.annotations** approach. I assume I'm not the first trying to use plac to write a command-line tool that needs to be distributed with a setup.py file, and the standard way to do that, AFAIK, is via the `console_scripts` feature. Would you have any hints on how to do that? I do not know how plac.call() works, and how it passes the arguments received from sys.argv to the annotated main function... Your help would be very appreciated.

**Comment by micheles** [comment]
I think the function can have arguments, as soon as they have defaults. Try to add default for repo_url, commit_id (i.e. rupo_url='' and then inside the main you can add a logic such as printing an error if repo_url is not specified).

**Comment by copernico** [comment]
I will try that, but it would be a temporary workaround: ideally, one would certainly prefer to be able to express the mandatory nature of parameters by omitting defaults, and plug plac on top, without changes due to the technicalities of how different pieces of the interface with the comman line work.

I took a look at the command-line script that is generated when running setup.py install, and here's what it looks like:
```
#!/home/xxxxxxx/.local/share/virtualenvs/git_explorer-Hd8x6_Ku/bin/python3
# EASY-INSTALL-ENTRY-SCRIPT: 'git-explorer','console_scripts','git-explorer-similar'
__requires__ = 'git-explorer'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('git-explorer', 'console_scripts', 'git-explorer-similar')()
    )
```

Does this additional information allows you to give me some other tips?

**Comment by micheles** [comment]
I just fixed a bug affecting Jupyter notebooks. It could have fixed your issue too.

**Comment by mhucka** [comment]
I realize this is a closed issue, but for anyone who (like me) may come here from a Google search trying to figure out how to make a plac-based interface work with setuptools's `console_scripts`, here is what worked for me for a program that has a similar organization as the OP's. 

First, I had to write a very small interface function to my main program, which I put where my main is defined (`__main__.py`, which seems to be the equivalent of the OP's `find_similar.py`):

```
def console_scripts_main():
    plac.call(main)
```

(The new function can be named anything other than `main`.) Then, in `setup.py` (or in my case, `setup.cfg`, but the effect is the same), I used this function as the entry point – here, "handprint" comes from the fact that my package is named "handprint" (the equivalent of the OP's `git-explorer`, I guess):

```
handprint = handprint.__main__:console_scripts_main
```

That was all that was needed, amazingly.

My program's `main` had a somewhat different ordering of optional versus required arguments compared to the OP's, and I'm not sure if that makes a difference. (See [the actual code in GitHub](https://github.com/caltechlibrary/handprint/blob/master/handprint/__main__.py), in case it's useful.)
